### PR TITLE
change nccl version reporting

### PR DESCRIPTION
**Pull request by eqy**
#62295

Previously the packing and unpacking of the NCCL version "integer" was done to have parity with the upstream NCCL version encoding. However, there doesn't seem to be any place where this integer is directly compared with a version integer sourced from upstream NCCL, and syncing the encoding seems to be error-prone (e.g., a recent change where a special case was added for minor versions >= 10 https://github.com/NVIDIA/nccl/blob/7e515921295adaab72adf56ea71a0fafb0ecb5f3/src/nccl.h.in#L22).

This patch changes the reporting to return a tuple of version numbers instead (to preserve ease-of-use for comparisons) and tweaks the passing between C/Python to avoid the digit overflow problem.

CC @ngimel @mcarilli

### CODE REVIEW

File: torch/csrc/cuda/nccl.cpp [hunk 1]
```diff
@@ -329,7 +329,10 @@ bool is_available(TensorList tensors) {
 
 std::uint64_t version() {
 #if defined(NCCL_MAJOR)
-  return NCCL_MAJOR * 1000 + NCCL_MINOR * 100 + NCCL_PATCH;
+  constexpr std::uint64_t ver = (((uint64_t) NCCL_MAJOR) << 32) |
+                                (((uint64_t) NCCL_MINOR) << 16) |
+                                ((uint64_t) NCCL_PATCH);
+  return ver;
 #elif defined(USE_NCCL)
   return 1000;
```
Comment (mrshenli) [inline torch/csrc/cuda/nccl.cpp:337]: IIUC, this will be parsed as `0.0.1000` by `nccl.version`?

**Review by mrshenli — APPROVED** [review]
LGTM

File: torch/csrc/cuda/nccl.cpp (hunk 1 above)
Comment (eqy) [reply torch/csrc/cuda/nccl.cpp:337]: good catch, this should be `1 << 32` in that case!

**Comment by jerryzh168** [comment]
test https://github.com/pytorch/pytorch/runs/3296491626 started failing after this PR, should we revert this?

**Comment by eqy** [comment]
Is this related to this PR? IIRC NCCL isn't used on Windows

**Comment by ngimel** [comment]
This test is failing after reland #62937, please don't revert #62916

**Pull request closed without merging**